File: sim/arena/navigator.ts
import { Arena } from './arena'
import { Edge }  from './edge'
import { Node }  from './node'

export class Navigator {

  // Distance matrix
  private dist: number[][]

  // Next node index in path
  private next: number[][]

  // Next edge in path
  private edgeNext: Edge[][]

  constructor(private arena: Arena) {
    this.floydWarshall()
  }

  // Returns the distance between two nodes.
  distance(from: Node, to: Node): number {
    const fromIdx = from.id
    const toIdx = to.id
    return this.dist[fromIdx][toIdx]
  }

  // Returns a path of edges from -> to.  Returns empty array if from and to are
  // the same node.  Returns null if there is no path from -> to.
  path(from: Node, to: Node): Edge[] | null {
    if (from === to) {
      return []
    }
    const fromIdx = from.id
    const toIdx = to.id

    if (this.next[fromIdx][toIdx] === undefined) {
      return null
    }

    const path = []
    let curIdx = fromIdx
    let prevIdx = fromIdx
    while (curIdx !== toIdx) {
      prevIdx = curIdx
      curIdx = this.next[curIdx][toIdx]

      path.push(this.edgeNext[prevIdx][curIdx])
    }

    return path
  }

  // Returns the next edge to travel down on the path from -> to.  Null if there
  // is no path, or if from and to are the same node.
  nextEdge(from: Node, to: Node): Edge | undefined {
    if (from === to) {
      return undefined
    }

    const edgeID = this.next[from.id][to.id]
    if (edgeID === undefined) {
      return undefined
    }

    return this.edgeNext[from.id][edgeID]
  }

  private static initDistanceMatrix(nodes: Node[]): number[][] {
    const mat = new Array(nodes.length)
    for (let i = 0; i < nodes.length; i++) {
      // The distance between every pair of nodes begins at Infinity
      mat[i] = new Array(nodes.length).fill(Infinity)

      // Except that every node is 0 away from itself
      mat[i][i] = 0
    }
    return mat
  }

  private static emptyMatrix<T>(size: number): T[][] {
    const mat = new Array(size)
    for (let i = 0; i < size; i++) {
      mat[i] = new Array(size)
    }
    return mat
  }

  // All-pairs shortest path algorithm
  // https://en.wikipedia.org/wiki/Floyd%E2%80%93Warshall_algorithm
  // Based on: https://github.com/cytoscape/cytoscape.js/blob/master/src/collection/algorithms/floyd-warshall.js
  private floydWarshall(): void {
    const nodes = this.arena.nodes
    const edges = this.arena.edges

    const dist = Navigator.initDistanceMatrix(nodes)
    const next = Navigator.emptyMatrix<number>(nodes.length)
    const edgeNext = Navigator.emptyMatrix<Edge>(nodes.length)

    // Record the distance between each pair of nodes that have an edge
    // between them.
    for (let i = 0; i < edges.length; i++) {
      const edge = edges[i]
      const leftID = edge.left.id
      const rightID = edge.right.id

      if (dist[leftID][rightID] > edge.weight) {
        dist[leftID][rightID] = edge.weight
        next[leftID][rightID] = rightID
        edgeNext[leftID][rightID] = edge
      }

      if (dist[rightID][leftID] > edge.weight) {
        dist[rightID][leftID] = edge.weight
        next[rightID][leftID] = leftID
        edgeNext[rightID][leftID] = edge
      }
    }

    // Build shortest-path matrix
    for (let k = 0; k < nodes.length; k++) {
      for (let i = 0; i < nodes.length; i++) {
        for (let j = 0; j < nodes.length; j++) {
          if (dist[i][k] + dist[k][j] < dist[i][j]) {
            dist[i][j] = dist[i][k] + dist[k][j]
            next[i][j] = next[i][k]
          }
        }
      }
    }

    this.dist = dist
    this.next = next
    this.edgeNext = edgeNext
  }

}

File: sim/navigator.go
package sim

import (
	"math"
)

// Navigator determines the shortest path between two nodes in an arena.
type Navigator struct {
	arena *Arena

	// Distance matrix, counting weights.
	Dist [][]uint64

	// Distance matrix, counting edges.
	EdgeDist [][]uint64

	// Table for next node ID in path between two nodes.
	Next [][]uint64
}

// NewNavigator creates a new Navigator from an Arena.
func NewNavigator(p *Provider) *Navigator {
	arena := p.Arena()
	navigator := &Navigator{
		arena:    arena,
		Dist:     initDistanceMatrix(arena.Nodes),
		EdgeDist: initDistanceMatrix(arena.Nodes),
		Next:     emptyMatrix(len(arena.Nodes)),
	}

	// floydWarshall(navigator, p)
	floydWarshall2(navigator, p)

	return navigator
}

// Distance returns the total distance (counting edge weights) between two
// nodes.
func (n *Navigator) Distance(from, to *Node) uint64 {
	return n.Dist[from.ID][to.ID]
}

// EdgeDistance returns the total number of edges between two nodes.
func (n *Navigator) EdgeDistance(from, to *Node) uint64 {
	return n.EdgeDist[from.ID][to.ID]
}

// NextEdge returns the next edge to travel down on the path from -> to.
func (n *Navigator) NextEdge(from, to *Node) *Edge {
	if from == to {
		return nil
	}

	nextID := n.Next[to.ID][from.ID]

	for _, edge := range from.Edges {
		if edge.A.ID == nextID || edge.B.ID == nextID {
			return edge
		}
	}

	return nil
}

func emptyMatrix(size int) [][]uint64 {
	storage := make([]uint64, size*size)
	mat := make([][]uint64, size)

	for i := range mat {
		n := i * size
		mat[i] = storage[n : n+size]
	}

	return mat
}

func initDistanceMatrix(nodes []*Node) [][]uint64 {
	dist := emptyMatrix(len(nodes))

	for i := 0; i < len(nodes); i++ {
		for j := 0; j < len(nodes); j++ {
			// The distance between every pair of nodes begins at "infinity".
			dist[i][j] = math.MaxUint64
		}

		// Except that every node is 0 away from itself.
		dist[i][i] = 0
	}

	return dist
}

func floydWarshall(nav *Navigator, p *Provider) {
	size := len(nav.arena.Nodes)

	dist := nav.Dist
	next := nav.Next
	edist := nav.EdgeDist

	// Record the distance between each pair of nodes that have an edge between
	// them.
	for _, edge := range nav.arena.Edges {
		if dist[edge.A.ID][edge.B.ID] > edge.Weight {
			dist[edge.A.ID][edge.B.ID] = edge.Weight
			next[edge.A.ID][edge.B.ID] = edge.B.ID
			edist[edge.A.ID][edge.B.ID] = 1
		}

		if dist[edge.B.ID][edge.A.ID] > edge.Weight {
			dist[edge.B.ID][edge.A.ID] = edge.Weight
			next[edge.B.ID][edge.A.ID] = edge.A.ID
			edist[edge.B.ID][edge.A.ID] = 1
		}
	}

	step := size / 100
	for step*100 < size {
		step++
	}

	// Build shortest-path matrix
	for k := 0; k < size; k++ {
		for i := 0; i < size; i++ {
			dIK := dist[i][k]
			if dIK == math.MaxUint64 {
				continue
			}

			for j := 0; j < size; j++ {
				dKJ := dist[k][j]
				if dKJ == math.MaxUint64 {
					continue
				}

				ij := &dist[i][j]
				dIKJ := dIK + dKJ

				if *ij > dIKJ {
					*ij = dIKJ
					next[i][j] = next[i][k]
					edist[i][j] = edist[i][k] + edist[k][j]
				}

			}
		}

		if k%step == 0 {
			p.Logger().Printf("navigation table %.0f%% complete (%d of %d)",
				float64(k)/float64(size)*100, k, size)
		}
	}
}

func floydWarshall2(nav *Navigator, p *Provider) {
	size := len(nav.arena.Nodes)

	dist := nav.Dist
	next := nav.Next
	edist := nav.EdgeDist

	// Record the distance between each pair of nodes that have an edge between
	// them.
	for _, edge := range nav.arena.Edges {
		if dist[edge.A.ID][edge.B.ID] > edge.Weight {
			dist[edge.A.ID][edge.B.ID] = edge.Weight
			next[edge.A.ID][edge.B.ID] = edge.B.ID
			edist[edge.A.ID][edge.B.ID] = 1
		}

		if dist[edge.B.ID][edge.A.ID] > edge.Weight {
			dist[edge.B.ID][edge.A.ID] = edge.Weight
			next[edge.B.ID][edge.A.ID] = edge.A.ID
			edist[edge.B.ID][edge.A.ID] = 1
		}
	}

	step := size / 100
	for step*100 < size {
		step++
	}

	// Build shortest-path matrix
	for k := 0; k < size; k++ {
		for i := 0; i < k; i++ {
			dIK := dist[i][k]

			if dIK == math.MaxUint64 {
				continue
			}

			for j := 0; j <= i; j++ {
				dKJ := dist[j][k]
				if dKJ == math.MaxUint64 {
					continue
				}

				ij := &dist[i][j]
				dIKJ := dIK + dKJ

				if *ij > dIKJ {
					*ij = dIKJ
					next[i][j] = next[i][k]
					edist[i][j] = edist[i][k] + edist[j][k]

					dist[j][i] = dIKJ
					next[j][i] = next[i][k]
					edist[j][i] = edist[i][k] + edist[j][k]
				}
			}
		}

		for i := k; i < size; i++ {
			dKI := dist[k][i]
			if dKI == math.MaxUint64 {
				continue
			}

			for j := 0; j < k; j++ {
				dKJ := dist[j][k]
				if dKJ == math.MaxUint64 {
					continue
				}

				ij := &dist[i][j]
				dIKJ := dKI + dKJ

				if *ij > dIKJ {
					*ij = dIKJ
					next[i][j] = next[k][i]
					edist[i][j] = edist[k][i] + edist[j][k]

					dist[j][i] = dIKJ
					next[j][i] = next[k][i]
					edist[j][i] = edist[k][i] + edist[j][k]
				}
			}

			for j := k; j <= i; j++ {
				dKJ := dist[k][j]
				if dKJ == math.MaxUint64 {
					continue
				}

				ij := &dist[i][j]
				dIKJ := dKI + dKJ

				if *ij > dIKJ {
					*ij = dIKJ
					next[i][j] = next[k][i]
					edist[i][j] = edist[k][i] + edist[k][j]

					dist[j][i] = dIKJ
					next[j][i] = next[k][i]
					edist[j][i] = edist[k][i] + edist[k][j]
				}
			}
		}

		if k%step == 0 {
			p.Logger().Printf("navigation table %.0f%% complete (%d of %d)",
				float64(k)/float64(size)*100, k, size)
		}
	}
}
